File: features/steps/canvas.py
# type: ignore
from behave import *

# from pprint import pprint
# if output is needed, add `print("\n\n")` at the end
# import rayz.util as U
from rayz.canvas import *

@given('c ← canvas({width}, {height})')
def step_impl(context, width, height):
    new_canvas = Canvas(width=int(width), height=int(height))
    context.canvas = new_canvas

# THEN #######################################################################

@then('every pixel of c is color(0, 0, 0)')
def step_impl(context):
    for index, row in enumerate(context.canvas.grid):
        assert (row == [Color(red=0, green=0, blue=0) for _ in range(len(row))]) is True

@then('pixel_at(c, {w}, {h}) = red')
def step_impl(context, w, h):
        red = Color(red=1, green=0, blue=0)
        color = context.canvas.pixel_at(int(w), int(h))
        assert (color == red) is True

@then('lines 1-3 of ppm are')
def step_impl(context):
    lines = context.ppm.split("\n")
    assert (lines[0] == "P3") is True
    assert (lines[1] == "5 3") is True
    assert (lines[2] == "255") is True

@then('lines 4-6 of ppm are')
def step_impl(context):
    lines = context.ppm.split("\n")
    assert (lines[3] == "255 0 0 0 0 0 0 0 0 0 0 0 0 0 0") is True
    assert (lines[4] == "0 0 0 0 0 0 0 128 0 0 0 0 0 0 0") is True
    assert (lines[5] == "0 0 0 0 0 0 0 0 0 0 0 0 0 0 255") is True

@then('ppm ends with a newline character')
def step_impl(context):
    assert (context.ppm.endswith("\n")) is True

# @then(u'lines 4-7 of ppm are')
# def step_impl(context):
#     lines = context.ppm.split("\n")
#     assert (lines[3] == "255 204 153 255 204 153 255 204 153 255 204 153 255 204 153 255 204") is True
#     assert (lines[4] == "153 255 204 153 255 204 153 255 204 153 255 204 153") is True
#     assert (lines[5] == "255 204 153 255 204 153 255 204 153 255 204 153 255 204 153 255 204") is True
#     assert (lines[6] == "153 255 204 153 255 204 153 255 204 153 255 204 153") is True

# WHEN #######################################################################

@when('write_pixel(c, {w}, {h}, {color})')
def step_impl(context, w, h, color):
    match color:
        case 'red':
            red = Color(red=1, green=0, blue=0)
            context.canvas.write_pixel(int(w), int(h), red)
        case 'c1':
            context.canvas.write_pixel(int(w), int(h), context.c1)
        case 'c2':
            context.canvas.write_pixel(int(w), int(h), context.c2)
        case 'c3':
            context.canvas.write_pixel(int(w), int(h), context.c3)

@when('ppm ← canvas_to_ppm(c)')
def step_impl(context):
    context.ppm = context.canvas.to_ppm()

@when('every pixel of c is set to color({r}, {g}, {b})')
def step_impl(context, r, g, b):
    for index, row in enumerate(context.canvas.grid):
        assert (row == [Color(red=float(r), green=float(g), blue=float(b)) for _ in range(len(row))]) is True

File: features/step_definitions/canvas.rb
Given('canvas ← canvas\({int}, {int})') do |int, int2|
  @canvas = Rayz::Canvas.new(width: int, height: int2)
end

Then("canvas.width = {int}") do |int|
  assert_equal(@canvas.width, int)
end

Then("canvas.height = {int}") do |int|
  assert_equal(@canvas.height, int)
end

Then('every pixel of canvas is color\({float}, {float}, {float})') do |float, float2, float3|
  expected_color = Rayz::Color.new(red: float, green: float2, blue: float3)

  (0..@canvas.height - 1).to_a.reverse_each do |r|
    (0..@canvas.width - 1).each do |c|
      assert_equal(@canvas.pixels[r][c], expected_color)
    end
  end
end

Given('red ← color\({float}, {float}, {float})') do |float, float2, float3|
  @red = Rayz::Color.new(red: float, green: float2, blue: float3)
end

When('write_pixel\(canvas, {int}, {int}, red)') do |int, int2|
  @canvas.write_pixel(col: int, row: int2, color: @red)
end

When('write_pixel\(canvas, {int}, {int}, color1)') do |int, int2|
  @canvas.write_pixel(col: int, row: int2, color: @color1)
end

When('write_pixel\(canvas, {int}, {int}, color2)') do |int, int2|
  @canvas.write_pixel(col: int, row: int2, color: @color2)
end

When('write_pixel\(canvas, {int}, {int}, color3)') do |int, int2|
  @canvas.write_pixel(col: int, row: int2, color: @color3)
end

Then('pixel_at\(canvas, {int}, {int}) = red') do |int, int2|
  color = @canvas.pixel_at(row: int, col: int2)
  assert_equal(color, @red)
end

When('ppm ← canvas_to_ppm\(canvas)') do
  @ppm = @canvas.to_ppm
end

Then("lines 1-3 of ppm are") do |doc_string|
  ppm_lines = @ppm.split("\n")[0..2]
  doc_lines = doc_string.split("\n")
  (0..ppm_lines.length - 1).each do |idx|
    assert_equal(ppm_lines[idx], doc_lines[idx])
  end
end

Then("lines 4-6 of ppm are") do |doc_string|
  low = 3
  ppm_lines = @ppm.split("\n")[low..]
  doc_lines = doc_string.split("\n")
  (0..ppm_lines.length - 1).each do |idx|
    assert_equal(ppm_lines[idx], doc_lines[idx])
  end
end

Then("ppm ends with a newline character") do
  assert_equal(@ppm.end_with?("\n"), true)
end
